File: src/test/java/dev/jiaming/ai_interview/coach/CoachRagContextServiceTests.java
package dev.jiaming.ai_interview.coach;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import org.junit.jupiter.api.Test;

import dev.jiaming.ai_interview.document.DocumentChunk;
import dev.jiaming.ai_interview.document.DocumentSourceType;
import dev.jiaming.ai_interview.document.ResolvedDocument;
import dev.jiaming.ai_interview.rag.RagIndexingService;
import dev.jiaming.ai_interview.rag.RagContextSnippet;
import dev.jiaming.ai_interview.rag.RagDocumentIndexHandle;
import dev.jiaming.ai_interview.rag.RagRetrievalService;
import dev.jiaming.ai_interview.resume.SectionAwareTextChunker;

class CoachRagContextServiceTests {

	@Test
	void shortDocumentsUseEveryPersistedChunkWithoutEmbedding() {
		RagIndexingService indexingService = mock(RagIndexingService.class);
		RagRetrievalService retrievalService = mock(RagRetrievalService.class);
		CoachRagContextService service = new CoachRagContextService(
			new SectionAwareTextChunker(),
			indexingService,
			retrievalService,
			new SimpleMeterRegistry()
		);
		ResolvedDocument resume = document(
			DocumentSourceType.RESUME,
			"resume-hash",
			"short resume",
			List.of(
				new DocumentChunk(0, "Experience", "Built an API", "resume:experience:0"),
				new DocumentChunk(1, "Projects", "Built an AI coach", "resume:projects:1")
			)
		);
		ResolvedDocument jobDescription = document(
			DocumentSourceType.JOB_DESCRIPTION,
			"jd-hash",
			"short job description",
			List.of(new DocumentChunk(0, "Requirements", "Java required", "job_description:requirements:0"))
		);

		CoachRagContext context = service.assessmentContext(new CoachAnalysisInput(
			resume,
			Optional.of(jobDescription),
			"Backend Engineer",
			"Mid-level"
		));

		assertThat(context.vectorBacked()).isFalse();
		assertThat(context.context()).contains("Built an API", "Built an AI coach", "Java required");
		assertThat(context.sourceContextIds()).containsExactly(
			"resume:experience:0",
			"resume:projects:1",
			"job_description:requirements:0"
		);
		verifyNoInteractions(indexingService, retrievalService);
	}

	@Test
	void localDocumentContextIsReservedWhenAnotherDocumentUsesVectorRetrieval() {
		RagIndexingService indexingService = mock(RagIndexingService.class);
		RagRetrievalService retrievalService = mock(RagRetrievalService.class);
		CoachRagContextService service = new CoachRagContextService(
			new SectionAwareTextChunker(),
			indexingService,
			retrievalService,
			new SimpleMeterRegistry()
		);
		ResolvedDocument resume = document(
			DocumentSourceType.RESUME,
			"resume-hash",
			"R".repeat(6_100),
			List.of(new DocumentChunk(0, "Experience", "Resume experience", "resume:experience:0"))
		);
		ResolvedDocument jobDescription = document(
			DocumentSourceType.JOB_DESCRIPTION,
			"jd-hash",
			"Java and distributed systems are required",
			List.of(new DocumentChunk(
				0,
				"Requirements",
				"Java and distributed systems are required",
				"job_description:requirements:0"
			))
		);
		RagDocumentIndexHandle handle = new RagDocumentIndexHandle(UUID.randomUUID(), 1L);
		when(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(handle));
		when(indexingService.ensureIndexed(jobDescription)).thenReturn(Optional.empty());
		when(retrievalService.retrieve(anyString(), any(RagDocumentIndexHandle.class), eq(6))).thenReturn(
			java.util.stream.IntStream.range(0, 14)
				.mapToObj(index -> new RagContextSnippet(
					"resume-vector-" + index,
					"Resume vector context " + index,
					java.util.Map.of(
						"contextId", "resume:experience:" + index,
						"sourceType", "resume",
						"section", "Experience"
					),
					0.9
				))
				.toList()
		);

		CoachRagContext context = service.assessmentContext(new CoachAnalysisInput(
			resume,
			Optional.of(jobDescription),
			"Backend Engineer",
			"Mid-level"
		));

		assertThat(context.vectorBacked()).isTrue();
		assertThat(context.context()).contains("Java and distributed systems are required");
		assertThat(context.sourceContextIds()).contains("job_description:requirements:0");
	}

	@Test
	void retrievalFailureUsesLocalContextAndIsNotReportedAsVectorBacked() {
		RagIndexingService indexingService = mock(RagIndexingService.class);
		RagRetrievalService retrievalService = mock(RagRetrievalService.class);
		CoachRagContextService service = new CoachRagContextService(
			new SectionAwareTextChunker(),
			indexingService,
			retrievalService,
			new SimpleMeterRegistry()
		);
		ResolvedDocument resume = document(
			DocumentSourceType.RESUME,
			"resume-hash",
			"R".repeat(6_100),
			List.of(new DocumentChunk(0, "Experience", "Resume experience", "resume:experience:0"))
		);
		RagDocumentIndexHandle handle = new RagDocumentIndexHandle(UUID.randomUUID(), 1L);
		when(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(handle));
		when(retrievalService.retrieve(anyString(), any(RagDocumentIndexHandle.class), eq(6)))
			.thenThrow(new IllegalStateException("vector store unavailable"));

		CoachRagContext context = service.assessmentContext(new CoachAnalysisInput(
			resume,
			Optional.empty(),
			"Backend Engineer",
			"Mid-level"
		));

		assertThat(context.vectorBacked()).isFalse();
		assertThat(context.context()).contains("Resume experience");
		assertThat(context.sourceContextIds()).containsExactly("resume:experience:0");
	}

	@Test
	void reservesJobDescriptionEvidenceAndRestoresOriginalChunkText() {
		RagIndexingService indexingService = mock(RagIndexingService.class);
		RagRetrievalService retrievalService = mock(RagRetrievalService.class);
		CoachRagContextService service = new CoachRagContextService(
			new SectionAwareTextChunker(), indexingService, retrievalService, new SimpleMeterRegistry()
		);
		ResolvedDocument resume = document(
			DocumentSourceType.RESUME, "resume-hash", "R".repeat(6_100),
			java.util.stream.IntStream.range(0, 6).mapToObj(index -> new DocumentChunk(
				index, "Experience", "Resume evidence " + index, "resume:experience:" + index
			)).toList()
		);
		ResolvedDocument jobDescription = document(
			DocumentSourceType.JOB_DESCRIPTION, "jd-hash", "J".repeat(200),
			java.util.stream.IntStream.range(0, 3).mapToObj(index -> new DocumentChunk(
				index, "Requirements", "Original JD requirement " + index,
				"job_description:requirements:" + index
			)).toList()
		);
		RagDocumentIndexHandle resumeIndex = new RagDocumentIndexHandle(UUID.randomUUID(), 1L);
		RagDocumentIndexHandle jobIndex = new RagDocumentIndexHandle(UUID.randomUUID(), 1L);
		when(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(resumeIndex));
		when(indexingService.ensureIndexed(jobDescription)).thenReturn(Optional.of(jobIndex));
		when(retrievalService.retrieve(anyString(), eq(resumeIndex), eq(6))).thenReturn(java.util.stream.IntStream.range(0, 6)
			.mapToObj(index -> snippet("resume:experience:" + index, "resume", "Experience", index,
				"Section: Experience\\nResume evidence " + index)).toList());
		when(retrievalService.retrieve(anyString(), eq(jobIndex), eq(4))).thenReturn(java.util.stream.IntStream.range(0, 3)
			.mapToObj(index -> snippet("job_description:requirements:" + index, "job_description", "Requirements", index,
				"Section: Requirements\\nOriginal JD requirement " + index)).toList());

		CoachRagContext context = service.assessmentContext(new CoachAnalysisInput(
			resume, Optional.of(jobDescription), "Backend Engineer", "Mid-level"
		));

		assertThat(context.sourceContextIds()).contains(
			"job_description:requirements:0", "job_description:requirements:1", "job_description:requirements:2"
		);
		assertThat(context.context()).contains("Original JD requirement 0").doesNotContain("Section: Requirements");
		verify(retrievalService, times(4)).retrieve(anyString(), eq(resumeIndex), eq(6));
		verify(retrievalService, times(4)).retrieve(anyString(), eq(jobIndex), eq(4));
	}

	@Test
	void usesContextIdAsDeterministicRrfTieBreaker() {
		RagIndexingService indexingService = mock(RagIndexingService.class);
		RagRetrievalService retrievalService = mock(RagRetrievalService.class);
		CoachRagContextService service = new CoachRagContextService(
			new SectionAwareTextChunker(), indexingService, retrievalService, new SimpleMeterRegistry()
		);
		ResolvedDocument resume = document(
			DocumentSourceType.RESUME, "resume-hash", "R".repeat(6_100),
			List.of(
				new DocumentChunk(0, "Experience", "Alpha evidence", "resume:experience:0"),
				new DocumentChunk(1, "Experience", "Beta evidence", "resume:experience:1")
			)
		);
		RagDocumentIndexHandle index = new RagDocumentIndexHandle(UUID.randomUUID(), 1L);
		when(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(index));
		AtomicInteger calls = new AtomicInteger();
		when(retrievalService.retrieve(anyString(), eq(index), eq(6))).thenAnswer(invocation -> {
			boolean alphaFirst = calls.getAndIncrement() % 2 == 0;
			RagContextSnippet alpha = snippet("resume:experience:0", "resume", "Experience", 0, "Alpha evidence");
			RagContextSnippet beta = snippet("resume:experience:1", "resume", "Experience", 1, "Beta evidence");
			return alphaFirst ? List.of(alpha, beta) : List.of(beta, alpha);
		});

		CoachRagContext context = service.assessmentContext(new CoachAnalysisInput(
			resume, Optional.empty(), "Backend Engineer", "Mid-level"
		));

		assertThat(context.sourceContextIds()).startsWith("resume:experience:0", "resume:experience:1");
	}

	private RagContextSnippet snippet(String contextId, String source, String section, int index, String content) {
		return new RagContextSnippet(
			"vector-" + contextId, content,
			java.util.Map.of("contextId", contextId, "sourceType", source, "section", section, "chunkIndex", index),
			0.9
		);
	}

	private ResolvedDocument document(
		DocumentSourceType sourceType,
		String hash,
		String text,
		List<DocumentChunk> chunks
	) {
		return new ResolvedDocument(sourceType, UUID.randomUUID(), hash, text, chunks);
	}
}

File: src/test/kotlin/dev/jiaming/ai_interview/coach/CoachRagContextServiceTests.kt
package dev.jiaming.ai_interview.coach

import dev.jiaming.ai_interview.document.DocumentChunk
import dev.jiaming.ai_interview.document.DocumentSourceType
import dev.jiaming.ai_interview.document.ResolvedDocument
import dev.jiaming.ai_interview.rag.RagContextSnippet
import dev.jiaming.ai_interview.rag.RagDocumentIndexHandle
import dev.jiaming.ai_interview.rag.RagIndexingService
import dev.jiaming.ai_interview.rag.RagRetrievalService
import dev.jiaming.ai_interview.resume.SectionAwareTextChunker
import io.micrometer.core.instrument.simple.SimpleMeterRegistry
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito
import java.util.Optional
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

class CoachRagContextServiceTests {

    @Test
    fun shortDocumentsUseEveryPersistedChunkWithoutEmbedding() {
        val indexingService = Mockito.mock(RagIndexingService::class.java)
        val retrievalService = Mockito.mock(RagRetrievalService::class.java)
        val service = service(indexingService, retrievalService)
        val resume = document(DocumentSourceType.RESUME, "resume-hash", "short resume", listOf(
            DocumentChunk(0, "Experience", "Built an API", "resume:experience:0"),
            DocumentChunk(1, "Projects", "Built an AI coach", "resume:projects:1")
        ))
        val jobDescription = document(DocumentSourceType.JOB_DESCRIPTION, "jd-hash", "short job description", listOf(
            DocumentChunk(0, "Requirements", "Java required", "job_description:requirements:0")
        ))

        val context = service.assessmentContext(CoachAnalysisInput(resume, Optional.of(jobDescription), "Backend Engineer", "Mid-level"))

        assertThat(context.vectorBacked()).isFalse()
        assertThat(context.context()).contains("Built an API", "Built an AI coach", "Java required")
        assertThat(context.sourceContextIds()).containsExactly(
            "resume:experience:0", "resume:projects:1", "job_description:requirements:0"
        )
        Mockito.verifyNoInteractions(indexingService, retrievalService)
    }

    @Test
    fun localDocumentContextIsReservedWhenAnotherDocumentUsesVectorRetrieval() {
        val indexingService = Mockito.mock(RagIndexingService::class.java)
        val retrievalService = Mockito.mock(RagRetrievalService::class.java)
        val service = service(indexingService, retrievalService)
        val resume = document(DocumentSourceType.RESUME, "resume-hash", "R".repeat(6_100), listOf(
            DocumentChunk(0, "Experience", "Resume experience", "resume:experience:0")
        ))
        val jobDescription = document(DocumentSourceType.JOB_DESCRIPTION, "jd-hash", "Java and distributed systems are required", listOf(
            DocumentChunk(0, "Requirements", "Java and distributed systems are required", "job_description:requirements:0")
        ))
        val handle = RagDocumentIndexHandle(UUID.randomUUID(), 1L)
        Mockito.`when`(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(handle))
        Mockito.`when`(indexingService.ensureIndexed(jobDescription)).thenReturn(Optional.empty())
        Mockito.`when`(retrievalService.retrieve(anyString(), any(RagDocumentIndexHandle::class.java), eq(6))).thenReturn(
            (0 until 14).map { index -> RagContextSnippet(
                "resume-vector-$index", "Resume vector context $index",
                mapOf("contextId" to "resume:experience:$index", "sourceType" to "resume", "section" to "Experience"), 0.9
            ) }
        )

        val context = service.assessmentContext(CoachAnalysisInput(resume, Optional.of(jobDescription), "Backend Engineer", "Mid-level"))

        assertThat(context.vectorBacked()).isTrue()
        assertThat(context.context()).contains("Java and distributed systems are required")
        assertThat(context.sourceContextIds()).contains("job_description:requirements:0")
    }

    @Test
    fun retrievalFailureUsesLocalContextAndIsNotReportedAsVectorBacked() {
        val indexingService = Mockito.mock(RagIndexingService::class.java)
        val retrievalService = Mockito.mock(RagRetrievalService::class.java)
        val service = service(indexingService, retrievalService)
        val resume = document(DocumentSourceType.RESUME, "resume-hash", "R".repeat(6_100), listOf(
            DocumentChunk(0, "Experience", "Resume experience", "resume:experience:0")
        ))
        val handle = RagDocumentIndexHandle(UUID.randomUUID(), 1L)
        Mockito.`when`(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(handle))
        Mockito.`when`(retrievalService.retrieve(anyString(), any(RagDocumentIndexHandle::class.java), eq(6)))
            .thenThrow(IllegalStateException("vector store unavailable"))

        val context = service.assessmentContext(CoachAnalysisInput(resume, Optional.empty(), "Backend Engineer", "Mid-level"))

        assertThat(context.vectorBacked()).isFalse()
        assertThat(context.context()).contains("Resume experience")
        assertThat(context.sourceContextIds()).containsExactly("resume:experience:0")
    }

    @Test
    fun reservesJobDescriptionEvidenceAndRestoresOriginalChunkText() {
        val indexingService = Mockito.mock(RagIndexingService::class.java)
        val retrievalService = Mockito.mock(RagRetrievalService::class.java)
        val service = service(indexingService, retrievalService)
        val resume = document(DocumentSourceType.RESUME, "resume-hash", "R".repeat(6_100),
            (0 until 6).map { index -> DocumentChunk(index, "Experience", "Resume evidence $index", "resume:experience:$index") })
        val jobDescription = document(DocumentSourceType.JOB_DESCRIPTION, "jd-hash", "J".repeat(200),
            (0 until 3).map { index -> DocumentChunk(index, "Requirements", "Original JD requirement $index", "job_description:requirements:$index") })
        val resumeIndex = RagDocumentIndexHandle(UUID.randomUUID(), 1L)
        val jobIndex = RagDocumentIndexHandle(UUID.randomUUID(), 1L)
        Mockito.`when`(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(resumeIndex))
        Mockito.`when`(indexingService.ensureIndexed(jobDescription)).thenReturn(Optional.of(jobIndex))
        Mockito.`when`(retrievalService.retrieve(anyString(), eq(resumeIndex), eq(6))).thenReturn(
            (0 until 6).map { index -> snippet("resume:experience:$index", "resume", "Experience", index, "Section: Experience\nResume evidence $index") })
        Mockito.`when`(retrievalService.retrieve(anyString(), eq(jobIndex), eq(4))).thenReturn(
            (0 until 3).map { index -> snippet("job_description:requirements:$index", "job_description", "Requirements", index, "Section: Requirements\nOriginal JD requirement $index") })

        val context = service.assessmentContext(CoachAnalysisInput(resume, Optional.of(jobDescription), "Backend Engineer", "Mid-level"))

        assertThat(context.sourceContextIds()).contains(
            "job_description:requirements:0", "job_description:requirements:1", "job_description:requirements:2"
        )
        assertThat(context.context()).contains("Original JD requirement 0").doesNotContain("Section: Requirements")
        Mockito.verify(retrievalService, Mockito.times(4)).retrieve(anyString(), eq(resumeIndex), eq(6))
        Mockito.verify(retrievalService, Mockito.times(4)).retrieve(anyString(), eq(jobIndex), eq(4))
    }

    @Test
    fun usesContextIdAsDeterministicRrfTieBreaker() {
        val indexingService = Mockito.mock(RagIndexingService::class.java)
        val retrievalService = Mockito.mock(RagRetrievalService::class.java)
        val service = service(indexingService, retrievalService)
        val resume = document(DocumentSourceType.RESUME, "resume-hash", "R".repeat(6_100), listOf(
            DocumentChunk(0, "Experience", "Alpha evidence", "resume:experience:0"),
            DocumentChunk(1, "Experience", "Beta evidence", "resume:experience:1")
        ))
        val index = RagDocumentIndexHandle(UUID.randomUUID(), 1L)
        Mockito.`when`(indexingService.ensureIndexed(resume)).thenReturn(Optional.of(index))
        val calls = AtomicInteger()
        Mockito.`when`(retrievalService.retrieve(anyString(), eq(index), eq(6))).thenAnswer {
            val alphaFirst = calls.getAndIncrement() % 2 == 0
            val alpha = snippet("resume:experience:0", "resume", "Experience", 0, "Alpha evidence")
            val beta = snippet("resume:experience:1", "resume", "Experience", 1, "Beta evidence")
            if (alphaFirst) listOf(alpha, beta) else listOf(beta, alpha)
        }

        val context = service.assessmentContext(CoachAnalysisInput(resume, Optional.empty(), "Backend Engineer", "Mid-level"))

        assertThat(context.sourceContextIds()).startsWith("resume:experience:0", "resume:experience:1")
    }

    private fun service(indexingService: RagIndexingService, retrievalService: RagRetrievalService) = CoachRagContextService(
        SectionAwareTextChunker(), indexingService, retrievalService, SimpleMeterRegistry()
    )

    private fun snippet(contextId: String, source: String, section: String, index: Int, content: String) =
        RagContextSnippet("vector-$contextId", content, mapOf(
            "contextId" to contextId, "sourceType" to source, "section" to section, "chunkIndex" to index
        ), 0.9)

    private fun document(sourceType: DocumentSourceType, hash: String, text: String, chunks: List<DocumentChunk>) =
        ResolvedDocument(sourceType, UUID.randomUUID(), hash, text, chunks)
}
